cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerTwillRunnable.java

```diff
@@ -0,0 +1,179 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * The {@link TwillRunnable} for running {@link ArtifactLocalizerService}.
+ *
+ * This runnable will run as a sidecar container for {@link io.cdap.cdap.internal.app.worker.TaskWorkerTwillRunnable}
+ */
+public class ArtifactLocalizerTwillRunnable extends AbstractTwillRunnable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactLocalizerTwillRunnable.class);
+
+  private ArtifactLocalizerService artifactLocalizerService;
+  private LogAppenderInitializer logAppenderInitializer;
+
+  public ArtifactLocalizerTwillRunnable(String cConfFileName, String hConfFileName) {
+    super(ImmutableMap.of("cConf", cConfFileName, "hConf", hConfFileName));
+  }
+
+  @VisibleForTesting
+  static Injector createInjector(CConfiguration cConf, Configuration hConf) {
+    List<Module> modules = new ArrayList<>();
+
+    modules.add(new ConfigModule(cConf, hConf));
+    modules.add(new IOModule());
+
+    // If MasterEnvironment is not available, assuming it is the old hadoop stack with ZK, Kafka
+    MasterEnvironment masterEnv = MasterEnvironments.getMasterEnvironment();
+
+    if (masterEnv == null) {
+      modules.add(new ZKClientModule());
+      modules.add(new ZKDiscoveryModule());
+      modules.add(new KafkaClientModule());
+      modules.add(new KafkaLogAppenderModule());
+    } else {
+      modules.add(new AbstractModule() {
+        @Override
+        protected void configure() {
+          bind(DiscoveryService.class)
+            .toProvider(new SupplierProviderBridge<>(masterEnv.getDiscoveryServiceSupplier()));
+          bind(DiscoveryServiceClient.class)
+            .toProvider(new SupplierProviderBridge<>(masterEnv.getDiscoveryServiceClientSupplier()));
+        }
+      });
+      modules.add(new RemoteLogAppenderModule());
+    }
+
+    return Guice.createInjector(modules);
+  }
+
+  @Override
+  public void initialize(TwillContext context) {
+    super.initialize(context);
+
+    try {
+      doInitialize();
+    } catch (Exception e) {
+      Throwables.propagateIfPossible(e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void run() {
+    CompletableFuture<Service.State> future = new CompletableFuture<>();
+    artifactLocalizerService.addListener(new ServiceListenerAdapter() {
+      @Override
+      public void terminated(Service.State from) {
+        future.complete(from);
+      }
+
+      @Override
+      public void failed(Service.State from, Throwable failure) {
+        future.completeExceptionally(failure);
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+
+    LOG.debug("Starting artifact localizer");
+    artifactLocalizerService.start();
+
+    try {
+      Uninterruptibles.getUninterruptibly(future);
+      LOG.debug("Artifact localizer stopped");
```

Comment:
Too many logs. You've already logged inside the stop implementation of the service